File: articles/calculus-for-systems-modeling/articles/definite-integrals-and-total-change/fortran/definite_integral_audit.f90
program definite_integral_audit
  implicit none
  real(8), dimension(9) :: times
  real(8) :: signed_accumulation, absolute_accumulation, dt, r0, r1
  integer :: i
  times = (/0.0d0,0.5d0,1.0d0,1.5d0,2.0d0,2.5d0,3.0d0,3.5d0,4.0d0/)
  signed_accumulation = 0.0d0
  absolute_accumulation = 0.0d0

  do i=1,8
    dt = times(i+1) - times(i)
    r0 = net_rate(times(i))
    r1 = net_rate(times(i+1))
    signed_accumulation = signed_accumulation + 0.5d0 * (r0 + r1) * dt
    absolute_accumulation = absolute_accumulation + 0.5d0 * (abs(r0) + abs(r1)) * dt
  end do

  print '(A)', 'interval_start interval_end signed_accumulation absolute_accumulation'
  print '(F8.3,1X,F8.3,1X,F14.6,1X,F14.6)', times(1), times(9), signed_accumulation, absolute_accumulation

contains
  real(8) function net_rate(t)
    real(8), intent(in) :: t
    net_rate = 4.0d0 * sin(t / 2.0d0) + 1.0d0
  end function
end program

File: articles/calculus-for-systems-modeling/articles/definite-integrals-and-total-change/c/definite_integral_audit.c
#include <math.h>
#include <stdio.h>
#include <stdlib.h>

double net_rate(double t){ return 4.0 * sin(t / 2.0) + 1.0; }

int main(void){
  double times[] = {0,0.5,1,1.5,2,2.5,3,3.5,4};
  double signed_accumulation = 0.0;
  double absolute_accumulation = 0.0;

  for(int i=0;i<8;i++){
    double dt = times[i+1] - times[i];
    double r0 = net_rate(times[i]);
    double r1 = net_rate(times[i+1]);
    signed_accumulation += 0.5 * (r0 + r1) * dt;
    absolute_accumulation += 0.5 * (fabs(r0) + fabs(r1)) * dt;
  }

  printf("interval_start,interval_end,method,signed_accumulation,absolute_accumulation\n");
  printf("%.6f,%.6f,trapezoidal approximation,%.12f,%.12f\n", times[0], times[8], signed_accumulation, absolute_accumulation);
  return EXIT_SUCCESS;
}
